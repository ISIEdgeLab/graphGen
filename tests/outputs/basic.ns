# Autogenerated by nsGen.py - Mon Jan 22 21:34:14 2018 UTC
set ns [new Simulator]
source tb_compat.tcl

set magi_str "sudo python /proj/edgect/magi/current/magi_bootstrap.py -fp /proj/edgect/magi/current/"

set click_str "/proj/edgect/exp_scripts/startClickDPDK.sh basic"

set my_start "touch /tmp/my_start"

# Enclaves

# Enclave 1
set server1 [$ns node]
tb-set-hardware $server1 MicroCloud
set traf11 [$ns node]
tb-set-hardware $traf11 MicroCloud
set traf12 [$ns node]
tb-set-hardware $traf12 MicroCloud
set traf13 [$ns node]
tb-set-hardware $traf13 MicroCloud
set traf14 [$ns node]
tb-set-hardware $traf14 MicroCloud
set traf15 [$ns node]
tb-set-hardware $traf15 MicroCloud
set traf16 [$ns node]
tb-set-hardware $traf16 MicroCloud
set traf17 [$ns node]
tb-set-hardware $traf17 MicroCloud
set traf18 [$ns node]
tb-set-hardware $traf18 MicroCloud
set ct1 [$ns node]
tb-set-node-os $ct1 Ubuntu1404-64-STD
tb-set-hardware $ct1 MicroCloud
set crypto1 [$ns node]
tb-set-node-os ${crypto1} Ubuntu1404-64-STD
tb-set-hardware ${crypto1} MicroCloud

# Enclave 2
set server2 [$ns node]
tb-set-hardware $server2 MicroCloud
set traf21 [$ns node]
tb-set-hardware $traf21 MicroCloud
set traf22 [$ns node]
tb-set-hardware $traf22 MicroCloud
set traf23 [$ns node]
tb-set-hardware $traf23 MicroCloud
set traf24 [$ns node]
tb-set-hardware $traf24 MicroCloud
set traf25 [$ns node]
tb-set-hardware $traf25 MicroCloud
set traf26 [$ns node]
tb-set-hardware $traf26 MicroCloud
set traf27 [$ns node]
tb-set-hardware $traf27 MicroCloud
set traf28 [$ns node]
tb-set-hardware $traf28 MicroCloud
set ct2 [$ns node]
tb-set-node-os $ct2 Ubuntu1404-64-STD
tb-set-hardware $ct2 MicroCloud
set crypto2 [$ns node]
tb-set-node-os ${crypto2} Ubuntu1404-64-STD
tb-set-hardware ${crypto2} MicroCloud

set vrouter [$ns node]
tb-set-node-os $vrouter Ubuntu1604-STD
tb-set-hardware $vrouter dl380g3
set control [$ns node]
tb-set-node-os $control Ubuntu1404-64-STD
tb-set-hardware $control MicroCloud

# External Nodes

# Lans
set lan1 [$ns make-lan "ct1 server1 traf11 traf12 traf13 traf14 traf15 traf16 traf17 traf18" 1Gb 0ms]
set lan2 [$ns make-lan "ct2 server2 traf21 traf22 traf23 traf24 traf25 traf26 traf27 traf28" 1Gb 0ms]

# Internal Links
set link1 [$ns duplex-link $ct1 ${crypto1} 1Gb 0ms DropTail]
set link2 [$ns duplex-link $ct2 ${crypto2} 1Gb 0ms DropTail]

# Egress Links
set elink1 [$ns duplex-link $crypto1 $vrouter 1Gb 0ms DropTail]
set elink2 [$ns duplex-link $crypto2 $vrouter 1Gb 0ms DropTail]

tb-set-vlink-emulation "vlan"
tb-set-multiplexed ${elink1} 1
tb-set-noshaping ${elink1} 1
tb-set-multiplexed ${elink2} 1
tb-set-noshaping ${elink2} 1

# External Node Links

# IPS

# IPs for Enclave 1
tb-set-ip-lan $traf11 $lan1 10.1.1.1
tb-set-ip-lan $traf12 $lan1 10.1.1.2
tb-set-ip-lan $traf13 $lan1 10.1.1.3
tb-set-ip-lan $traf14 $lan1 10.1.1.4
tb-set-ip-lan $traf15 $lan1 10.1.1.5
tb-set-ip-lan $traf16 $lan1 10.1.1.6
tb-set-ip-lan $traf17 $lan1 10.1.1.7
tb-set-ip-lan $traf18 $lan1 10.1.1.8
tb-set-ip-lan $server1 $lan1 10.1.1.9
tb-set-ip-lan $ct1 $lan1 10.1.1.100
tb-set-ip-link $ct1 ${link1} 10.1.2.1
tb-set-ip-link ${crypto1} ${link1} 10.1.2.2

# IPs for Enclave 2
tb-set-ip-lan $traf21 $lan2 10.2.1.1
tb-set-ip-lan $traf22 $lan2 10.2.1.2
tb-set-ip-lan $traf23 $lan2 10.2.1.3
tb-set-ip-lan $traf24 $lan2 10.2.1.4
tb-set-ip-lan $traf25 $lan2 10.2.1.5
tb-set-ip-lan $traf26 $lan2 10.2.1.6
tb-set-ip-lan $traf27 $lan2 10.2.1.7
tb-set-ip-lan $traf28 $lan2 10.2.1.8
tb-set-ip-lan $server2 $lan2 10.2.1.9
tb-set-ip-lan $ct2 $lan2 10.2.1.100
tb-set-ip-link $ct2 ${link2} 10.2.2.1
tb-set-ip-link ${crypto2} ${link2} 10.2.2.2

# Egress link IPS
tb-set-ip-link ${crypto1} ${elink1} 10.1.10.1
tb-set-ip-link $vrouter ${elink1} 10.1.10.2
tb-set-ip-link ${crypto2} ${elink2} 10.2.10.1
tb-set-ip-link $vrouter ${elink2} 10.2.10.2

# External Node Link IPS 

# Write Tee Nodes

# Container Partitioning
tb-add-node-attribute $traf11 containers:partition 1
tb-add-node-attribute $traf12 containers:partition 1
tb-add-node-attribute $traf13 containers:partition 1
tb-add-node-attribute $traf14 containers:partition 1
tb-add-node-attribute $traf15 containers:partition 1
tb-add-node-attribute $traf16 containers:partition 1
tb-add-node-attribute $traf17 containers:partition 1
tb-add-node-attribute $traf18 containers:partition 1
tb-add-node-attribute $server1 containers:partition 1
tb-add-node-attribute $traf21 containers:partition 2
tb-add-node-attribute $traf22 containers:partition 2
tb-add-node-attribute $traf23 containers:partition 2
tb-add-node-attribute $traf24 containers:partition 2
tb-add-node-attribute $traf25 containers:partition 2
tb-add-node-attribute $traf26 containers:partition 2
tb-add-node-attribute $traf27 containers:partition 2
tb-add-node-attribute $traf28 containers:partition 2
tb-add-node-attribute $server2 containers:partition 2

tb-add-node-attribute $ct1 containers:partition 3
tb-add-node-attribute ${crypto1} containers:partition 4
tb-add-node-attribute $ct2 containers:partition 5
tb-add-node-attribute ${crypto2} containers:partition 6

tb-add-node-attribute $vrouter containers:partition 7
tb-add-node-attribute $control containers:partition 8

# Embed Physical Nodes
tb-add-node-attribute $ct1 containers:node_type embedded_pnode
tb-add-node-attribute ${crypto1} containers:node_type embedded_pnode
tb-add-node-attribute $ct2 containers:node_type embedded_pnode
tb-add-node-attribute ${crypto2} containers:node_type embedded_pnode

tb-add-node-attribute $vrouter containers:node_type embedded_pnode
tb-add-node-attribute $control containers:node_type embedded_pnode

# Start Commands

tb-set-node-startcmd $traf11 "$my_start; $magi_str"
tb-set-node-startcmd $traf12 "$my_start; $magi_str"
tb-set-node-startcmd $traf13 "$my_start; $magi_str"
tb-set-node-startcmd $traf14 "$my_start; $magi_str"
tb-set-node-startcmd $traf15 "$my_start; $magi_str"
tb-set-node-startcmd $traf16 "$my_start; $magi_str"
tb-set-node-startcmd $traf17 "$my_start; $magi_str"
tb-set-node-startcmd $traf18 "$my_start; $magi_str"
tb-set-node-startcmd $server1 "$my_start; $magi_str"
tb-set-node-startcmd $ct1 "$my_start; $magi_str"
tb-set-node-startcmd ${crypto1} "$my_start; $magi_str"

tb-set-node-startcmd $traf21 "$my_start; $magi_str"
tb-set-node-startcmd $traf22 "$my_start; $magi_str"
tb-set-node-startcmd $traf23 "$my_start; $magi_str"
tb-set-node-startcmd $traf24 "$my_start; $magi_str"
tb-set-node-startcmd $traf25 "$my_start; $magi_str"
tb-set-node-startcmd $traf26 "$my_start; $magi_str"
tb-set-node-startcmd $traf27 "$my_start; $magi_str"
tb-set-node-startcmd $traf28 "$my_start; $magi_str"
tb-set-node-startcmd $server2 "$my_start; $magi_str"
tb-set-node-startcmd $ct2 "$my_start; $magi_str"
tb-set-node-startcmd ${crypto2} "$my_start; $magi_str"

tb-set-node-startcmd $vrouter "$click_str; $my_start; $magi_str"
tb-set-node-startcmd $control "sudo python /proj/edgect/exp_scripts/fixHosts.py; $my_start; $magi_str"

# Epilogue
$ns rtproto Static
$ns run
