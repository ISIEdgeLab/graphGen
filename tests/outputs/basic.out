// Auto generated Click Template using GraphGen/clickGen.py at Mon Jan 22 21:34:14 2018 UTC
// Command line: {'arp': False, 'cryptoHardware': 'MicroCloud', 'ns_file': 'basic.ns', 'clientHardware': 'MicroCloud', 'writeRoutes': False, 'useDPDK': True, 'useCrypto': True, 'ct_os': 'Ubuntu1404-64-STD', 'draw_output': None, 'ctHardware': 'MicroCloud', 'useCodel': True, 'useContainers': True, 'loss': 0.0, 'os': 'Ubuntu1404-64-STD', 'clickHardware': 'dl380g3', 'delay': '0ms', 'writePaths': '', 'numServers': 1, 'bw': '1Gbps', 'numClients': 8, 'routes': None, 'output': 'basic.out', 'inConstraints': False, 'startCmd': '', 'infile': 'basic.xnet'} 


// Routers
router1 :: RadixIPLookup(${if1_ip} 2,
                         ${if1_16} ${if1_gw} 0,
                         ${if2_16} 1,
                         10.100.150.2 1,
                         10.100.150.1 2);

router2 :: RadixIPLookup(${if2_ip} 2,
                         ${if1_16} 0,
                         ${if2_16} ${if2_gw} 1,
                         10.100.150.1 0,
                         10.100.150.2 2);


// Packet Classifiers
c1 :: Classifier(12/0800, 12/0806 20/0001, 12/0806 20/0002, -);
c2 :: Classifier(12/0800, 12/0806 20/0001, 12/0806 20/0002, -);
chost :: Classifier(12/0800, 12/0806 20/0001, 12/0806 20/0002, -);

// VLAN Multiplexing
vlanmux :: VlanSwitch(VLAN ${vlan1}, VLAN ${vlan2});

// Packet Arrival
$DPDKArrival
vlanmux[0] -> c1;
vlanmux[1] -> c2;
FromHost(fake0) -> chost;

// Packet Departure
$DPDKDeparture
out1 :: VLANEncap(${vlan1}) -> ${out_if1};
out2 :: VLANEncap(${vlan2}) -> ${out_if2};

// Handle ARP
arpt :: Tee(3);

c1[1] -> ar1 :: ARPResponder(${if1_ip} ${if1_eth}) -> out1;
arpq1 :: ARPQuerier(${if1_ip}, ${if1_eth}) -> out1;
c1[2] -> arpt;
arpt[0] -> [1]arpq1;

c2[1] -> ar2 :: ARPResponder(${if2_ip} ${if2_eth}) -> out2;
arpq2 :: ARPQuerier(${if2_ip}, ${if2_eth}) -> out2;
c2[2] -> arpt;
arpt[1] -> [1]arpq2;

chost[1] -> c1;
chost[2] -> arpt;

// Link Traffic Shaping
link_1_2_queue :: ThreadSafeQueue(1000);
link_1_2_bw :: LinkUnqueue(0ms, 1Gbps);
link_1_2_loss :: RandomSample(DROP 0.0);
link_2_1_queue :: ThreadSafeQueue(1000);
link_2_1_bw :: LinkUnqueue(0ms, 1Gbps);
link_2_1_loss :: RandomSample(DROP 0.0);

// Decrement TTL and send time exceeded replies
r1ttl_2 :: DecIPTTL;
r1ttl_2[1] -> ICMPError(10.100.150.1, timeexceeded) -> router1;
r2ttl_1 :: DecIPTTL;
r2ttl_1[1] -> ICMPError(10.100.150.2, timeexceeded) -> router2;
r1ttl_out_e1 :: DecIPTTL;
r1ttl_out_e1[1] -> ICMPError(10.100.150.1, timeexceeded) -> router1;
r2ttl_out_e2 :: DecIPTTL;
r2ttl_out_e2[1] -> ICMPError(10.100.150.2, timeexceeded) -> router2;

// Links
router1[1] -> r1ttl_2 -> SetTimestamp(FIRST true) -> link_1_2_queue -> CoDel() -> link_1_2_loss -> link_1_2_bw -> router2
router2[0] -> r2ttl_1 -> SetTimestamp(FIRST true) -> link_2_1_queue -> CoDel() -> link_2_1_loss -> link_2_1_bw -> router1

 // Teed Inputs and Outputs

 // Input from Teed interfaces is discarded

// Output Chains

// Send IP Packets to Routers
c1[0] -> Strip(14) -> CheckIPHeader(0) -> router1;
c2[0] -> Strip(14) -> CheckIPHeader(0) -> router2;
chost[0] -> Strip(14) -> CheckIPHeader(0) -> router1;

// Send out packets to Interfaces
router1[0] -> r1ttl_out_e1 -> arpq1;
router2[1] -> r2ttl_out_e2 -> arpq2;

// Local Delivery
toh :: ToHost(fake0);

router1[2] -> EtherEncap(0x0800, 1:1:1:1:1:1, 2:2:2:2:2:2) -> toh;
router2[2] -> EtherEncap(0x0800, 1:1:1:1:1:1, 2:2:2:2:2:2) -> toh;
arpt[2] -> toh;

// Unknown packets to their death
c1[3] -> Print("${if1} Non IP") -> Discard;
c2[3] -> Print("${if2} Non IP") -> Discard;
chost[3] -> Print("Host Non IP") -> Discard;
